shocker.c

```diff
@@ -1,60 +1,161 @@
 #define _GNU_SOURCE
-#include <sys/wait.h> // waitpid
+#include <stdlib.h>
+#include <stdbool.h>
 #include <sched.h> // clone, all the clone Flags
 #include <unistd.h> // execvp, sethostname
 #include <string.h> // strlen
 #include <stdio.h> // puts, printf
 #include <sys/mount.h> // mount
+#include <sys/stat.h> // mkdir
+#include <sys/types.h> // Flags for sys stuff
+#include <sys/mman.h>
+#include <sys/wait.h> // waitpid
 
-#define STACK_SIZE (1024 * 1024)
+#define STACK_SIZE (1024 * 1024)    /* Stack size for cloned child */
 
 //   docker run image <cmd> <params>
 //./shocker run       <cmd> <params>
 
-static char child_stack[STACK_SIZE]; /* Stack size for cloned child */
-
 int child(void *arg) {
-  sethostname("Container", 10);
-  chdir("/root/container");
-  chroot("/root/container"); // this is the main part of hiding
-  mount("proc", "/proc", "proc", 0, NULL); // proc is a pseudo-fs, used by many tools
+  if (sethostname("Container", 10) == -1) {
+    perror("sethostname failed with: ");
+    goto fail_sethostname;
+  }
+
+  if (chdir("/root/container") == -1) {
+    perror("sethostname failed with: ");
+    goto fail_chdir;
+  }
+
+  if (chroot("/root/container") == -1) {
+    perror("sethostname failed with: ");
+    goto fail_chroot;
+  }
+
+  if (mount("proc", "/proc", "proc", 0, NULL) == -1) {
+    perror("mount failed with: ");
+    goto fail_mount;
+  }
+
   char **shell = (char **)arg;
-  execvp(shell[0], shell);
-  return 0;
+
+  if (execvp(shell[0], shell) == -1) {
+    perror("execvp failed with: ");
+    goto fail_execvp;
+  }
+
+  if (umount("/root/container/proc") == -1) {
+    perror("umount failed with: ");
+    goto fail_umount;
+  }
+
+  exit(EXIT_SUCCESS);
+
+  fail_umount:
+  fail_execvp:
+    umount("/root/container/proc");
+  fail_mount:
+  fail_chroot:
+  fail_chdir:
+  fail_sethostname:
+    exit(EXIT_FAILURE);
+}
+
+static bool intWriter(char* path, int num) {
+  FILE *fp = fopen(path, "w");
+
+  if (fp == NULL) {
+    fprintf(stderr, "fopen %s ", path);
+    perror("failed with: ");
```

Comment:
not a big deal but you can use `strerror` to extract the errno string which you can then pass into your `fprintf` statement. This produces the error in a single line instead of 2.

Reply:
actually a tiny bit more of an issue then I though because you're actively omitting the `\n` to make every go on the same line. That can be problematic if you have multiple threads or things printout and can lead to interleaved outputs.

Not a big deal in this code base but just mentioning it.

Reply:
Worth considering though.